unlock-app/src/hooks/useMultipleRecipient.ts

```diff
@@ -0,0 +1,193 @@
+  const config: any = useContext(ConfigContext)
+  const [retryBulkAction, setRetryBulkAction] = useState(true)
+  const { maxRecipients = 1, locks } = paywallConfig ?? {}
+  const lock = lockAddress ? locks?.[lockAddress] : {}
+
+  const normalizeRecipients = (ignoreRecipients?: User[]) => {
+    if (!lockAddress) return
+
+    const listWithoutExcluded = recipientsList().filter(
+      ({ resolvedAddress }) => {
+        return ignoreRecipients && ignoreRecipients?.length > 0
+          ? !ignoreRecipients
+              .map(({ userAddress }) => userAddress)
+              .includes(resolvedAddress)
+          : true
+      }
+    )
+
+    const payload: RecipientPayload = {
+      users: listWithoutExcluded.map(({ resolvedAddress, metadata = {} }) => {
+        return {
+          userAddress: resolvedAddress,
+          metadata,
+          lockAddress,
+        }
+      }),
+    }
+    return payload
+  }
+
+  const clear = () => {
+    setRecipients(new Set([]))
+  }
+
+  const addAccountToList = () => {
+    addRecipientItem(account)
+  }
+
+  const purchaseBulk = async () => {
+    if (!lockAddress) return
+    const owners = recipientsList().map(
+      ({ resolvedAddress }) => resolvedAddress
+    )
+
+    await purchaseMultipleKeys({
+      walletService,
+      lockAddress,
+      owners,
+    })
+  }
+
+  useEffect(() => {
+    const activeMultiple = +maxRecipients > 1
+    setHasMultipleRecipients(activeMultiple)
+    if (activeMultiple) {
+      addAccountToList()
+    }
+  }, [])
+
+  const recipientsList = (): RecipientItem[] => {
+    return Array.from(recipients)
+  }
+
+  const canAddUser = () => {
+    return recipients.size < maxRecipients
+  }
+
+  const getAddressAndValidation = async (recipient: string) => {
+    const address = await getAddressForName(recipient)
+    return {
+      valid: address?.length > 0 ?? false,
+      address,
+    }
+  }
+
+  const submitBulkRecipients = async (ignoreRecipients?: User[]) => {
+    if (!lock?.network) return
+    const url = `${config.services.storage.host}/v2/api/metadata/${lock?.network}/users`
+    const opts = {
+      method: 'POST',
+      body: JSON.stringify(normalizeRecipients(ignoreRecipients)),
+      headers: {
+        'content-type': 'application/json',
+      },
+    }
+    const response = await fetch(url, opts)
+    return response
+  }
+
+  const submit = async (ignoreRecipients?: User[], callback?: () => void) => {
+    const res = await submitBulkRecipients(ignoreRecipients)
+    if (res) {
+      const result: any = await res.json()
+      const errorCodes = [500, 401, 404, 409]
+      console.log(result, res)
+      if (errorCodes.includes(res.status)) {
+        ToastHelper.error(result?.message ?? 'Ops, something went wrong')
+        if (retryBulkAction) {
+          ToastHelper.success('Re-try bulk submit')
+          setRetryBulkAction(false)
+          submit(result?.users ?? [], callback)
+        }
+      } else {
```

Comment:
I would handle 409 separately because the error message should be `Metadata for x,y,z user already exists therefore it won't be added`. 

There should be maximum retry because otherwise, you will end up DDOSing the server.

Reply:
regardings the maximum retry, i'll stop from retrying by setting `retryBulkAction=true`, in line `138`, you have other suggestions?

Reply:
I was thinking of using number so you can control how many times it retries. Make sense?

Reply:
@searchableguy  Oh nice! so let me change it, and trigger the counter only when the user clicks the button